DESIGN.md

```diff
@@ -0,0 +1,55 @@
+# Design
+
+This document covers the design thoughts behind build images for speeding up Travis CI and a proposal for how to implement things going forward.
+
+## Background
+
+Travis CI currently uses standard Ubuntu containers to build curl. There are lots and lots of entries in the curl build matrix and currently a single job takes a long time to complete.
+
+Part of this is due to downloading and compiling dependencies (e.g. brotli, wolfssl). There's also elapsed time taken in downloading and installing packages for compilation.
+
+Using Docker containers, we can create a build image whereby the dependencies are already downloaded and installed, and compiled where necessary. Theoretically it also means developers can have the same build environment as the CI jobs, which is good for reproducibility.
+
+## Design thoughts
+
+- The build images should be based on Ubuntu Trusty, as Travis CI currently use Ubuntu Trusty as their base image.
```

Comment:
I think it is reasonable that the first step is to keep things similar (eg. using Ubuntu Trusty) and increment from there.

Reply:
Yep. I'm kinda inclined to use Xenial given it's the majority image of the build jobs now. @bagder; no idea if you want to weigh in here about which distribution you'd prefer to build on by default?

Reply:
I don't have any strong opinion either way. For our travis CI builds we strive to switch to as recent versions as possible when we can.

Reply:
Ok. With this approach we should theoretically be able to run the newest Ubuntu version if desired. For now, let's do Xenial, but with an eye to easily replacing this in future.